diff --git a/db/version_edit.h b/db/version_edit.h
index a069893..99ac1f3 100644
--- a/db/version_edit.h
+++ b/db/version_edit.h
@@ -84,7 +84,7 @@ class VersionEdit {
  private:
   friend class VersionSet;
 
-  typedef std::set< std::pair<int, uint64_t> > DeletedFileSet;
+  USED_BY_NESTED_FRIEND2(typedef std::set< std::pair<int, uint64_t> > DeletedFileSet)
 
   std::string comparator_;
   uint64_t log_number_;
@@ -97,9 +97,9 @@ class VersionEdit {
   bool has_next_file_number_;
   bool has_last_sequence_;
 
-  std::vector< std::pair<int, InternalKey> > compact_pointers_;
-  DeletedFileSet deleted_files_;
-  std::vector< std::pair<int, FileMetaData> > new_files_;
+  USED_BY_NESTED_FRIEND2(std::vector< std::pair<int, InternalKey> > compact_pointers_)
+  USED_BY_NESTED_FRIEND(DeletedFileSet deleted_files_)
+  USED_BY_NESTED_FRIEND2(std::vector< std::pair<int, FileMetaData> > new_files_)
 };
 
 }
diff --git a/db/version_set.h b/db/version_set.h
index 2dbd948..74bbe1f 100644
--- a/db/version_set.h
+++ b/db/version_set.h
@@ -101,7 +101,7 @@ class Version {
   int refs_;                    // Number of live refs to this version
 
   // List of files per level
-  std::vector<FileMetaData*> files_[config::kNumLevels];
+  USED_BY_NESTED_FRIEND(std::vector<FileMetaData*> files_[config::kNumLevels])
 
   // Next file to compact based on seek stats.
   FileMetaData* file_to_compact_;
diff --git a/port/port_posix.h b/port/port_posix.h
index 9666391..fd4d840 100644
--- a/port/port_posix.h
+++ b/port/port_posix.h
@@ -43,6 +43,17 @@
 #define fdatasync fsync
 #endif
 
+// Some compilers do not provide access to nested classes of a declared friend class
+// Defining PUBLIC_NESTED_FRIEND_ACCESS will cause those declarations to be made
+// public as a workaround
+#if defined(OS_MACOSX) || defined(OS_SOLARIS)
+#define USED_BY_NESTED_FRIEND(a) public: a; private:
+#define USED_BY_NESTED_FRIEND2(a,b) public: a,b; private:
+#else
+#define USED_BY_NESTED_FRIEND(a) a;
+#define USED_BY_NESTED_FRIEND2(a,b) a,b;
+#endif
+
 namespace leveldb {
 namespace port {
 

